### ISSUE TITLE
Plan creation wizard rework

### ISSUE THREAD

**Issue by sjDCC** [DMPRoadmap/roadmap#2337]
Several users / unis have reported confusion over the plan creation wizard. It's not always intuitive to know how to get institutional templates. Perhaps we need to adjust the layout of this wizard so institutional templates are immediately visible beneath the org box (once selected) and it shows as an either/or to select a uni template or a funder one?

Dutch users have asked to discuss this at the next user group as they have a slightly different position emerging. ZonMw and NWO are allowing researchers to use uni templates when they are approved. This means DMPonline will need to change the logic for this use case to present relevant options. We can configure it as an option that's switched off for others.

Thoughts from US examples welcome @mariapraetzellis

**Comment by briri** [comment]
We've also had requests to allow a one-click plan creation type feature for organizational templates.

**Comment by sjDCC** [comment]
The main preference to come out of the user group was having basic filters for orgs and funders which then bring up a dropdown list of all possible options. Admin users would like to configure the names and order of that list to align with local policy e.g. use the TU Delft template

![Picture1](https://user-images.githubusercontent.com/6169940/79761616-9a0a2a80-8319-11ea-98bb-bb6310852688.png)

**Comment by magdalenadrafiova** [comment]
relates to: https://github.com/DMPRoadmap/roadmap/issues/2835

**Comment by mariapraetzellis** [comment]
Adding a recent comment from a user: 

> When creating a DMP, selecting a template is gated by the primary funding organization. It might be nice to have an option to use an alternative organization's template (e.g., the DOT template is not accessible if I list FAA as my primary research organization).

**Comment by magdalenadrafiova** [comment]
Not sure whether we discuss the same here @mariapraetzellis (might be) but we got suggestion that when a plan is created and the funder is selected the funder template will appear in the list of templates the researcher can choose. rather than automatically be presented with the funder template. In Netherlands the funders often approve institutional templates, and this is quite clear from their naming conventions but the institutional template would not appear the second funder template is selected/ Therefore having a list of templates where researchers could select one would be better way to go forward

**Comment by pherterich** [comment]
Linking another set of issues that should be considered as part of this work: #3088 #2886

**Comment by johnpinto1** [comment]
I have started on this. I see it is quite complex, so may take some time to understand functionality and then change it. As I need to understand current functionality to understand what may be feasible.

**Comment by johnpinto1** [comment]
> Adding a recent comment from a user:
> 
> > When creating a DMP, selecting a template is gated by the primary funding organization. It might be nice to have an option to use an alternative organization's template (e.g., the DOT template is not accessible if I list FAA as my primary research organization).

@mariapraetzellis a question regarding the user's suggestion. So in the quote cited is the user wanting to see all publicly available templates from other orgs too?

**Comment by pherterich** [comment]
> @mariapraetzellis a question regarding the user's suggestion. So in the quote cited is the user wanting to see all publicly available templates from other orgs too?

Yes, this is one of the issues we will need to address, some funders allow institutional templates to be used, so while the current approach is helpful to narrow it down to the right template, it can be too restrictive in cases. The initial mock-up suggests something like a few recommended templates but also the option to use "other" and search for more. We probably need to narrow things down to exactly what we want where, but this is the general challenge.

**Comment by johnpinto1** [comment]
@pherterich  All such input is helpful as I study and play around with the code.

**Comment by briri** [comment]
It may be worth having @benjaminfaure show us their plan creation workflow in more detail. They have a custom approach that I believe is more flexible than the baseline Roadmap approach

**Comment by johnpinto1** [comment]
@briri can you arrange it?

**Comment by mariapraetzellis** [comment]
@johnpinto1 and @pherterich, I'm excited to start working on this feature. A larger Roadmap team meeting to discuss approaches to implementing this would be helpful. I will propose some times via Slack.

**Comment by mariapraetzellis** [comment]
Just a note that we should be sure to make it easy for a user to generate a plan based on the generic/default template.

**Comment by johnpinto1** [comment]
@briri I think I have done the changes you suggested. Please double check https://github.com/DMPRoadmap/roadmap/commit/c645381aa213621640e07c9cdcd33dd2b23e8fd0.

**Comment by johnpinto1** [comment]
@briri  I have updated the code https://github.com/DMPRoadmap/roadmap/commit/fc5ac7e6f85db40f2d19e7b373365f39099fff4e . So the required asterisk appears beside template selection with associated tooltip and screen reader read only text.

![Selection_023](https://user-images.githubusercontent.com/8876215/176641831-74219c07-8a00-4e23-839c-050274c08603.png)

**Comment by nicolasfranck** [comment]
Shouldn't there be a place for administrative data? I mean: "primary research organisation" and "funder" look in the wizard like fields that do not have anything to do the selection of a template. A different layout would make that more clear:

### Choose your template:

  #### by primary research organisation 
  #### by funder

Some of our users are confused that about these fields..

**Comment by nicolasfranck** [comment]
Or:

#### Choose your template

`--Select your primary research organisation--` [ ] `No research organisation associated ..`
`--Select your funder--`[ ] `No funder associated..`
`--Select your template--`

All of them would have to be select boxes, and each of them is activated
by the select box above (or by checking the box next to it).

Every select box should have an empty value (with instructions),
so one must select a value, even if it is just one.
This way one knows what template was selected

**Comment by johnpinto1** [comment]
@nicolasfranck Thanks will incorporate one of suggestion when I revisit code.

**Comment by laurastandaert** [comment]
We have run into problems with users not selecting an appropriate template because they were trying to be as administratively correct as possible in the 'create plan' page, while they should actually be selecting the most appropriate template. This is a serious risk because this also means that they do not fill out the privacy list of questions (which is an institutional customisation to the template) we want them to answer.

We could think of 3 distinct functionalities where organisations and funders can be selected/entered in the DMP application
- **Purpose 1**: org to which the DMP belongs in db: important for visibility, org admins etc. This should not be modifiable
- **Purpose 2**: select a suitable template with tailored guidance
  - This is what you do in 'create plan' page and you want to make it as fool-proof as possible that users select the correct template with the instructions you want them to have. And you also want to make sure there is no confusion: this merely serves to select a template and guidance, not to have the most complete administrative information (see purpose 3)
- **Purpose 3**: administrative details about your project regarding organisation and funder
   - This is not limited to available templates and guidances, there can be many more affiliations or funding sources. Just as is now the case for the funder information on the 'project details': this does not influence the funder template as entered in 'create plan', nor the available customisations eg question specific guidance. 
   - A field similar to Funder could be added to Project Details to indicate the Research Organisation (or even call this 'affiliation' to not use the same terminology as 'organisation' in the application). 
   - In an ideal world, this information is standardized by selecting from RoR supplemented by an 'Other' field in which free text can be entered

**Comment by johnpinto1** [comment]
> Shouldn't there be a place for administrative data? I mean: "primary research organisation" and "funder" look in the wizard like fields that do not have anything to do the selection of a template. A different layout would make that more clear:
> ### Choose your template:
> #### by primary research organisation
> #### by funder
> 
> Some of our users are confused that about these fields..

@nicolasfranck I looked at your suggestion and realised that: this is not what the DCC & UC3 asked for. They want to the user to choose any template, with a default template being initially selected.

**Comment by johnpinto1** [comment]
> We have run into problems with users not selecting an appropriate template because they were trying to be as administratively correct as possible in the 'create plan' page, while they should actually be selecting the most appropriate template. This is a serious risk because this also means that they do not fill out the privacy list of questions (which is an institutional customisation to the template) we want them to answer.
> 
> We could think of 3 distinct functionalities where organisations and funders can be selected/entered in the DMP application
> 
>     * **Purpose 1**: org to which the DMP belongs in db: important for visibility, org admins etc. This should not be modifiable
> 
>     * **Purpose 2**: select a suitable template with tailored guidance
>       
>       * This is what you do in 'create plan' page and you want to make it as fool-proof as possible that users select the correct template with the instructions you want them to have. And you also want to make sure there is no confusion: this merely serves to select a template and guidance, not to have the most complete administrative information (see purpose 3)
> 
>     * **Purpose 3**: administrative details about your project regarding organisation and funder
>       
>       * This is not limited to available templates and guidances, there can be many more affiliations or funding sources. Just as is now the case for the funder information on the 'project details': this does not influence the funder template as entered in 'create plan', nor the available customisations eg question specific guidance.
>       * A field similar to Funder could be added to Project Details to indicate the Research Organisation (or even call this 'affiliation' to not use the same terminology as 'organisation' in the application).
>       * In an ideal world, this information is standardized by selecting from RoR supplemented by an 'Other' field in which free text can be entered

@laurastandaert as I pointed out to @nicolasfranck it looks like the requirements suggested by DCC & UC3 are different from your suggestions. I think we need @mariapraetzellis , @dsisu and @briri to come up with reconciles their requirement with your suggestions. I know of one of them being currently on holiday.

**Comment by johnpinto1** [comment]
@briri
@dsisu
@gjacob24
@laurastandaert 
@mariapraetzellis
@martaribeiro
@nicolasfranck

We have deployed code for the current iteration of the Plan wizard to
https://dmponline-dev.dcc.ac.uk/ 
(Please note because of expired certificate on site you will get warnings which you should ignore to create account and login.)
Then go to "Create Plan" page (this is the current page). On the page there is a button "Try New Plan Creation Wizard" which will take you to the new functionality for you to critique for forthcoming meeting.

**Warning to ignore:**
![Selection_049](https://user-images.githubusercontent.com/8876215/186659615-e8b5e253-baf4-4020-a8ab-a6cc9a461ad3.png)

**Current "Create Plan" page with button button "Try New Plan Creation Wizard" to link to new functionality**
![Selection_047](https://user-images.githubusercontent.com/8876215/186659657-f6be647d-e17b-4762-9fc5-21a7a24bee08.png)

**New Plan Creation page**
![Selection_048](https://user-images.githubusercontent.com/8876215/186659708-1e9e077f-20e0-4907-90d9-2a23e677ba66.png)

**Comment by briri** [comment]
@pengyin-shan @benjaminfaure see John's note above about the new create plan page.

**Comment by pengyin-shan** [comment]
> @pengyin-shan @benjaminfaure see John's note above about the new create plan page.

Great! unfortunately I cannot log in because I don't know the name and password...but the flow looks good to me. 
Just FYI that DMP Assistant *does not use Funder in workflow*, so we always commented out the part that shows 'Select the primary funding organization'. We allow users to input anything in the 'funder' field when creating a new plan. This field will be a '_unmanaged_' organization that saved to database without any other information

**Comment by mariapraetzellis** [comment]
@johnpinto1 I'm wondering if we need to add a typeahead in the Template field. DMPTool has many (+40) templates shared between all organizations, and some orgs have numerous institutional templates on top of this. So, a user will be presented with a long list of templates to scroll through. Adding a filter or typeahead could be a way to find the template they want quickly.

**Comment by johnpinto1** [comment]
@pengyin-shan You just create a new email account on server to login. Good luck!

**Comment by martaribeiro** [comment]
@johnpinto1 Do you think it would be possible to organise the dropdown content? 
I was thinking in something like:

Which DMP template would you like to use?
_______________________________________
Templates from Primary Research Organisation
	Template 1
	Template 2
	etc
_______________________________________
Templates from Funders
	Template 1
	Template 2
	etc
_______________________________________
All other templates
 	Template 1
	Template 2
	etc

**Comment by johnpinto1** [comment]
Will see I can incorporated both  suggestions by @martaribeiro & @mariapraetzellis

**Comment by johnpinto1** [comment]
@briri I have made the changes @mariapraetzellis requested using your suggestion of HTML5 datalist in commit https://github.com/DigitalCurationCentre/roadmap/commit/ebed027f9a626557b4fabb19f079d6bf48c15ba5
This will be deployed to Dev by @martaribeiro tomorrow. I will update.

**Comment by johnpinto1** [comment]
@briri I have made the changes @mariapraetzellis requested using your suggestion of HTML5 datalist in commit https://github.com/DMPRoadmap/roadmap/commit/adce18a61178c79f5e96b95c79b2d1dbf9a394b3

@martaribeiro has deployed the changes. 

Styling of the Template dropdown requires styling. I found I will need to use a gem datalist-css as otherwise datalist style depends on browser.

 @mariapraetzellis I hope this is what you requested

**Comment by johnpinto1** [comment]
@briri
@dsisu
@gjacob24
@laurastandaert
@mariapraetzellis
@martaribeiro
@nicolasfranck
@pengyin-shan 

I have updated the code which is deployed on the Dev server 
https://dmponline-dev.dcc.ac.uk/ 

Changes:
- Templates now can be searched in dropdown with typeahead.
- Templates in dropdown have been split into three categories: 
> Templates from Primary Research Organisations
Templates from Funders
Other Templates

Note I am using an HTML5 datalist for the Template dropdown. It cannot be easily styled and is browser dependent. I will talk to @briri about alternatives when he is back from his holidays.

**Comment by johnpinto1** [comment]
@laurastandaert found a bug in the New Plan Creation. You will get the following error if try to click on New Plan Creation button if you a new user or you are not an Org Admin. Let me know if you get error and I will temporarily make you Org Admin until I fix issue.
![Selection_027](https://user-images.githubusercontent.com/8876215/190219705-a9be8c84-60be-4410-af8b-0850a2ef3d48.png)

**Comment by johnpinto1** [comment]
I have a fix for bug above. So new users won't need to be made Org admins by me. It will be deployed this afternoon. So bear with us if you get error above.

**Comment by johnpinto1** [comment]
Fix now deployed. So all users should be able to create plans using New Plan page.

**Comment by gjacob24** [comment]
The plan creation branch will be temporarily unavailable until next week, as we need to do some testing on the roadmap one. We will let you know as soon as it is up again.

**Comment by mariapraetzellis** [comment]
Adding a link to the ideas [document](https://docs.google.com/document/d/1f-OFMQVfxfbL_ajG6XC75DqW6fzNmfPp/edit) shared by @nicolasfranck and Laura Standaert at UGent.

**Comment by johnpinto1** [comment]
A Swedish user has come up with some further suggestions:

> As for the plan creation wizard, there are some suggestions that might be worthwhile to consider.

> Why not change the current text on the front page of the plan creation wizard and replace “No research organisation associated with this plan or my research organisation is not listed” with a sentence like “I would like to add a research organisation that is not listed here”? In a similar way, one could replace “No funder associated with this plan or my funder is not listed” with a sentence like “I would like to add a funder that is not listed here”. That would probably sound a bit more welcoming and would likely be no big deal to change.

> As for research funders in Sweden, there is a long list of different funders that not only includes the Swedish Research Council and FORMAS but also many large and small private foundations. It is probably difficult to include them all, but it would not be difficult to invite a person who would like to write a data management in DMPonline to add their own funder and research organisation.

> That applies also to the page about “Project details”. There is no need to show researchers a red flag when they enter a research funder that is not on the default list of DMPonline. Instead a of a red message “The name you entered was not one of the listed suggestions!” it might perhaps be possible to write something like “You entered a funder that currently is not on the default list, which may be perfectly fine.”

> People in the UK are known for being very polite and why should DMPonline not speak the same kind of language? To add one more point about the page for “Project details”, it would likely be greatly appreciated, if there was a field that would allow people to add the research organisation that is responsible for the research project because this a very important information about a research project.  

> To add two more points that would make life a bit easier for people who use DMPonline, it would be very helpful, if the section about contributors would appear in the printed version of a data management plan because it would save us the effort to include this part as an own section into our template. In addition, we believe that it would make life easier for everybody, if the option “project detail cover sheet” would be preselected in the section “Download settings”. Not everybody looks at this point and as a result it can happen that people send in data management plans with no administrative information about their project. At least the second suggestion is likely a minor point that could be easily changed.

**Comment by johnpinto1** [comment]
@briri @benjaminfaure @dsisu@gjacob24 @laurastandaert @mariapraetzellis @martaribeiro @nicolasfranck
@pengyin-shan

A reworking of the Plan creation wizard based on a spec suggested by the folk at Ghent @laurastandaert  and @nicolasfranck. With coding help from @nicolasfranck to get me started as I was lost trying to implement the spec. I hope I have done it justice. All omissions are solely my responsibility. It is a start and a work in progress. The deployed code is at a new URL https://dmponline-test.dcc.ac.uk/ . Note you might need to register again or for first time. There is a certificate issue on server so ignore warnings to proceed.

**Comment by nicolasfranck** [comment]
Seems very good to me, but I'm just a programmer ;-)

Anyone else?

**Comment by johnpinto1** [comment]
@nicolasfranck Thanks. It appears our have contradictory requirements so we still have to sell this or change it to suit one group or the other.

**Comment by nicolasfranck** [comment]
@johnpinto1 sorry for taking so long, I was waiting for the end users to answer (to no avail). Do you have a list of the contradicting requirements/expectations of the end users?

**Comment by johnpinto1** [comment]
@nicolasfranck We have not created a list. But at an User meeting we had in November some folk liked the change, others said giving users a choice of templates will confuse some researchers.

**Comment by dsisu** [comment]
Also see issue #2835
